core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -177,8 +274,21 @@ public void registerSuggester(String key, Class<? extends Suggester> suggester)
         suggesters.registerExtension(key, suggester);
     }
 
-    public void registerFunctionScoreParser(Class<? extends ScoreFunctionParser> parser) {
-        functionScoreParsers.add(parser);
+    /**
+     * Register a new ScoreFunctionParser.
+     */
+    public void registerFunctionScoreParser(ScoreFunctionParser<?> parser) {
+        for (String name: parser.getNames()) {
+            Object oldValue = functionScoreParsers.putIfAbsent(name, parser);
+            if (oldValue != null) {
+                throw new IllegalArgumentException("Function score parser [" + oldValue + "] already registered for name [" + name + "]");
```

Comment:
They are both in the error message. I liked outputting the oldValue because it gives you some hint about what other component registered something with the same name.